// op-client/dist/src/entity/Abstract/Abstract.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const set_1 = __importDefault(require("keypather/set"));
const get_1 = __importDefault(require("keypather/get"));
class Abstract {
    constructor(init) {
        this.body = {
            id: undefined,
            _links: {
                self: undefined,
            },
            _embedded: {},
        };
        this._links = {};
        if (init) {
            if (typeof init === 'number') {
                this.id = init;
            }
            else if (typeof init === 'object' && init.hasOwnProperty('href')) {
                this.self = init;
            }
            else {
                this.merge(init);
                if (this.id && !this.self) {
                    this.id = this.id;
                }
                if (this.self && !this.id) {
                    this.self = this.self;
                }
            }
        }
    }
    get id() {
        return this.body.id;
    }
    set id(id) {
        this.body.id = id;
        this.body._links.self = this.body._links.self || { href: undefined };
        this.body._links.self.href = this.constructor.url + '/' + id;
        this.body._links.self.title = undefined;
    }
    merge(source) {
        const copy = JSON.parse(JSON.stringify(source));
        Object.assign(this.body, copy);
    }
    get self() {
        return this.body._links.self;
    }
    set self(value) {
        this.body._links.self = value;
        this.body.id = Number.parseInt(value.href.match(/\d+$/)[0]);
    }
    bodyScope(fieldPaths) {
        const result = {};
        fieldPaths.forEach(eachFieldPath => {
            const eachFieldValue = get_1.default(this.body, eachFieldPath);
            set_1.default(result, eachFieldPath, eachFieldValue);
        });
        return result;
    }
}
exports.default = Abstract;
Abstract.url = '/should_be_defined';
