// op-client/dist/src/entity/Abstract/Duration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const iso8601_duration_1 = require("iso8601-duration");
class Duration {
    constructor(parts) {
        this.years = 0;
        this.months = 0;
        this.days = 0;
        this.hours = 0;
        this.minutes = 0;
        this.seconds = 0;
        this.weeks = 0;
        if (parts) {
            Object.assign(this, parts);
        }
    }
    static parse(durationString) {
        const result = new Duration();
        Object.assign(result, iso8601_duration_1.parse(durationString));
        return result;
    }
    toString() {
        if (this.isEmpty()) {
            return 'PT0S';
        }
        else {
            return `P${this.years}Y${this.months}M${this.days}DT${this.hours}H${this.minutes}M${this.seconds}S`.replace(/(?<=[A-Z])0[A-Z]/g, '');
        }
    }
    isEmpty() {
        return iso8601_duration_1.toSeconds(this) === 0;
    }
    toSeconds() {
        return iso8601_duration_1.toSeconds(this);
    }
}
exports.default = Duration;
